airflow/providers/google/cloud/operators/bigquery.py

```diff
@@ -2039,6 +2039,149 @@ def execute(self, context) -> None:
+    )
+    template_fields_renderers = {"schema_fields": "json"}
+    ui_color = BigQueryUIColors.TABLE.value
+
+    @classmethod
+    def _patch_schema(cls, old_schema, new_schema):
+        # Updates the content of "old_schema" with
+        # the fields from new_schema. Makes changes
+        # in place and hence has no return value.
+        # Works recursively for sub-records
+
+        # Start by turning the schema list of fields into
+        # a dict keyed on the field name for both the old
+        # and the new schema
+        new_fields = {field["name"]: field for field in new_schema["fields"] if "name" in field}
+        old_fields = {field["name"]: field for field in old_schema["fields"]}
+
+        # Iterate over all new fields and update the
+        # old_schema dict
+        for field_key in new_fields.keys():
+            # Check if the field exists in the old_schema, if
+            # so change it
+            if field_key in old_fields:
+                old_field = old_fields[field_key]
+                new_field = new_fields[field_key]
+                # Check if recursion is needed
+                if "fields" in new_field:
+                    cls._patch_schema(old_field, new_field)
+                    del new_field["fields"]
+
+                # Do the update
+                old_field.update(new_field)
+
+            # Field didn't exist, add it as a new field
+            else:
+                old_schema["fields"].append(new_fields[field_key])
+
+    @apply_defaults
+    def __init__(
+        self,
+        *,
+        dataset_id: Optional[str] = None,
+        schema_fields: List[Dict[str, Any]],
+        table_id: Optional[str] = None,
+        project_id: Optional[str] = None,
+        gcp_conn_id: str = 'google_cloud_default',
+        delegate_to: Optional[str] = None,
+        impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+        **kwargs,
+    ) -> None:
+        self.schema_fields = schema_fields
+        self.table_id = table_id
+        self.dataset_id = dataset_id
+        self.project_id = project_id
+        self.gcp_conn_id = gcp_conn_id
+        self.delegate_to = delegate_to
+        self.impersonation_chain = impersonation_chain
+        super().__init__(**kwargs)
+
+    def execute(self, context):
+        bq_hook = BigQueryHook(
+            gcp_conn_id=self.gcp_conn_id,
+            delegate_to=self.delegate_to,
+            impersonation_chain=self.impersonation_chain,
+        )
+
+        table_schema = bq_hook.get_schema(
+            dataset_id=self.dataset_id, table_id=self.table_id, project_id=self.project_id
+        )
+
+        self._patch_schema(old_schema=table_schema, new_schema={"fields": self.schema_fields})
+
+        return bq_hook.update_table(
+            table_resource={"schema": table_schema},
+            fields=["schema"],
+            dataset_id=self.dataset_id,
+            table_id=self.table_id,
+            project_id=self.project_id,
+        )
```

Comment:
_patch_schema doesn't actually patch the schema on the table, so it doesn't interface with BigQuery, what it does is to take a schema object and alter the fields to be changed, after that the new schema is used as input to update_table where the actual call to the bigquery api is made. The leading underscore in the method indicates it is a private method not intended to be re-used outside this class, however it has to be broken out in a funciton for recursion to work properly

Reply:
do you think it better if I just define the function inside the execute scope? I mean, I could add all the logic from the Operator into a new hook - but then I don't see the point of separating hooks and operators, if operators are just wrappers

Reply:
Hooks are interfaces for Operators. The concept of [Operator](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#operators) and [Hook](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#hooks). Take a look at all others Operators in the BigQuery file.

IMHO, changing the function `_patch_schema` to BigqueryHook and create a public function called `update_schema` will be more useful in the future. Also change the name is important all other functions called `patch` are deprecated and now use `update` instead, you already use the update term in the documentation.
@turbaszek what do you think about it?

Reply:
I changed names to update instead of patch, however I do consider there to be a difference in that I don't require an entire new schema to be set but rather altering the existing fields. I have no strong feelings for naming though